qiskit/circuit/instruction.py

```diff
@@ -129,6 +129,52 @@ def __eq__(self, other):
 
         return True
 
+    def soft_compare(self, other: 'Instruction') -> bool:
+        """
+        Soft comparison between gates. Their names, number of qubits, and classical
+        bit numbers must match. The number of parameters must match. Each parameter
+        is compared. If one is a ParameterExpression then it is not taken int
+        account.
+
+        Args:
+            other (instruction): other instruction.
+
+        Returns:
+            bool: are self and other equal up to parameter expressions.
+        """
+        if self.name != other.name or \
+                other.num_qubits != other.num_qubits or \
+                other.num_clbits != other.num_clbits:
+            return False
+
+        for self_param, other_param in zip_longest(self.params, other.params):
+            try:
+                if self_param == other_param:
+                    continue
+            except ValueError:
+                pass
+
+            if not isinstance(self_param, ParameterExpression) and \
+                    not isinstance(other_param, ParameterExpression):
+                try:
+                    if numpy.shape(self_param) == numpy.shape(other_param) \
+                            and numpy.allclose(self_param, other_param, atol=_CUTOFF_PRECISION):
+                        continue
+                except TypeError:
+                    pass
+
+                try:
+                    if numpy.isclose(float(self_param), float(other_param), atol=_CUTOFF_PRECISION):
+                        continue
+                except TypeError:
+                    pass
+            else:
+                continue
```

Comment:
This is kind of hard to follow, is this the same?

```suggestion
            if isinstance(self_param, ParameterExpression) or \
                    isinstance(other_param, ParameterExpression):
                continue
            elif isinstance(self_param, numpy.array) and \
                    isinstance(other_param, numpy.array):
                if numpy.shape(self_param) == numpy.shape(other_param) \
                        and numpy.allclose(self_param, other_param, atol=_CUTOFF_PRECISION):
                    continue
            else:
                try:
                    if numpy.isclose(self_param, other_param, atol=_CUTOFF_PRECISION):
                        continue
                except TypeError:
                    pass
```

Reply:
I think this is the same. I'll commit the suggestion and run the tests.

Reply:
Right, it's a bit pseudo code-y. The correction is `isinstance(aa, np.ndarry)`

Reply:
I can update my comment

Reply:
Thanks that seems to work now. The soft_compare tests and template matching tests also pass.